src/previous.rs

```diff
@@ -0,0 +1,368 @@
+            + usize::from(discrepancy_budget % LARGE_EPOCH_LENGTH != 0))
+            * LARGE_EPOCH_LENGTH;
+        let mut total_jumps = LARGE_EPOCH_LENGTH;
+
+        while old_ratchet_large.large != recent.large {
+            let jump_count = old_ratchet_large.next_large_epoch();
+            total_jumps += jump_count;
+            if total_jumps > rounded_budget {
+                return Err(PreviousErr::BudgetExceeded);
+            }
+            iter.large_skips.push(old_ratchet_large.clone());
+        }
+
+        Ok(iter)
+    }
+
+    /// Returns the exact amount of ratchets between the old and recent ratchet.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use skip_ratchet::{Ratchet, PreviousIterator};
+    ///
+    /// let old_ratchet = Ratchet::from_rng(&mut rand::thread_rng());
+    /// let mut new_ratchet = old_ratchet.clone();
+    /// new_ratchet.inc_by(100_000);
+    ///
+    /// let iterator = PreviousIterator::new(&old_ratchet, &new_ratchet, 1_000_000_000).unwrap();
+    ///
+    /// assert_eq!(iterator.step_count(), 100_000);
+    /// ```
+    pub fn step_count(&self) -> usize {
+        let mut count = 0;
+
+        // The oldest ratchet in this iterator with the same large epoch as recent:
+        let (oldest_in_recent_epoch, from_large) = match self.large_skips.last() {
+            // the last large skip may or may not have the same large epoch
+            Some(last_large) if last_large.large == self.recent.large => (last_large, true),
+            // if not, we take the first in any of the other skip lists
+            _ => (
+                self.medium_skips
+                    .first()
+                    .or_else(|| self.small_skips.first())
+                    .unwrap_or(&self.recent),
+                false,
+            ),
+        };
+
+        // It's safe to compare combined counters, since they both are in the same large epoch
+        count += self.recent.combined_counter() - oldest_in_recent_epoch.combined_counter();
+
+        // Other than that we need to add all large skips
+        if let Some(first_large) = self.large_skips.first() {
+            count += (self.large_skips.len() - 1) * LARGE_EPOCH_LENGTH;
+            // If we didn't take the large from the large epoch, we need to add it back in here
+            if !from_large {
+                count += LARGE_EPOCH_LENGTH;
+            }
+            count += oldest_in_recent_epoch.combined_counter() - first_large.combined_counter();
+        }
+
+        count
+    }
+}
+
+impl Iterator for PreviousIterator {
+    type Item = Ratchet;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while !(self.large_skips.is_empty()
+            && self.medium_skips.is_empty()
+            && self.small_skips.is_empty())
+        {
+            if self.medium_skips.is_empty() && self.small_skips.is_empty() {
+                let old_ratchet_large = self.large_skips.pop().unwrap();
```

Comment:
`pop()` is fallible ?

Reply:
In theory yes, in practice no. It's pretty tricky, but the proptests should cover every possible case.

In this PR I only moved the file around, I didn't change this. If we find a better way to do this, I'd be super happy to merge :)

Reply:
If it's fallible it needs to be handled. If it should not fail then it should not be fallible.
Agreed we can handle it in follow-up PR but in a library PR everything should be panic free.